hpen.c

```diff
@@ -0,0 +1,491 @@
+#include <dev/evdev/evdev.h>
+
+#include "hid.h"
+#include "hidbus.h"
+
+#define	HPEN_DEBUG
+#define	HPEN_DEBUG_VAR	hpen_debug
+
+/* Check if debugging is enabled. */
+#ifdef HPEN_DEBUG_VAR
+#ifdef HPEN_DEBUG
+#define	DPRINTFN(n,fmt,...) do {					\
+	if ((HPEN_DEBUG_VAR) >= (n)) {					\
+		printf("%s: " fmt, __FUNCTION__ ,##__VA_ARGS__);	\
+	}								\
+} while (0)
+#define DPRINTF(...)    DPRINTFN(1, __VA_ARGS__)
+#else
+#define DPRINTF(...) do { } while (0)
+#define DPRINTFN(...) do { } while (0)
+#endif
+#endif
+
+#ifdef HPEN_DEBUG
+static int hpen_debug = 1;
+
+static SYSCTL_NODE(_hw_usb, OID_AUTO, hpen, CTLFLAG_RW, 0,
+		"Generic HID tablet");
+SYSCTL_INT(_hw_usb_hpen, OID_AUTO, debug, CTLFLAG_RWTUN,
+		&hpen_debug, 0, "Debug level");
+#endif
+
+
+#define	HPEN_NO_CODE	(ABS_MAX + 10)
+
+struct hpen_hid_map_item {
+	char		name[5];
+	int32_t 	usage;		/* HID usage */
+	uint32_t	code;		/* Evdev event code */
+	bool		required;	/* Required for Integrated Windows Pen tablets */
+};
+
+enum {
+	HPEN_X,
+	HPEN_Y,
+	HPEN_TIP_PRESSURE,
+	HPEN_X_TILT,
+	HPEN_Y_TILT,
+	HPEN_BATTERY_STRENGTH,
+	HPEN_N_USAGES_ABS,
+};
+
+static const struct hpen_hid_map_item hpen_hid_map_abs[HPEN_N_USAGES_ABS] = {
+	[HPEN_X] = {
+		.name = "X",
+		.usage = HID_USAGE2(HUP_GENERIC_DESKTOP, HUG_X),
+		.code = ABS_X,
+		.required = true,
+	},
+	[HPEN_Y] = {
+		.name = "Y",
+		.usage = HID_USAGE2(HUP_GENERIC_DESKTOP, HUG_Y),
+		.code = ABS_Y,
+		.required = true,
+	},
+	[HPEN_TIP_PRESSURE] = {
+		.name = "TPRS",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_TIP_PRESSURE),
+		.code = ABS_PRESSURE,
+		.required = true,
+	},
+	[HPEN_X_TILT] = {
+		.name = "XTLT",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_X_TILT),
+		.code = ABS_TILT_X,
+		.required = false,
+	},
+	[HPEN_Y_TILT] = {
+		.name = "YTLT",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_Y_TILT),
+		.code = ABS_TILT_Y,
+		.required = false,
+	},
+	[HPEN_BATTERY_STRENGTH] = {
+		.name = "BATT",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_BATTERY_STRENGTH),
+		.code = HPEN_NO_CODE, // TODO
+		.required = false,
+	},
+};
+
+enum {
+	HPEN_TIP_SWITCH,
+	HPEN_IN_RANGE,
+	HPEN_BARREL_SWITCH,
+	HPEN_INVERT,
+	HPEN_ERASER,
+	HPEN_N_USAGES_KEY,
+};
+
+static const struct hpen_hid_map_item hpen_hid_map_key[HPEN_N_USAGES_KEY] = {
+	[HPEN_TIP_SWITCH] = {
+		.name = "TIP",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_TIP_SWITCH),
+		.code = BTN_TOUCH,
+		.required = true,
+	},
+	[HPEN_IN_RANGE] = {
+		.name = "RNGE",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_IN_RANGE),
+		.code = BTN_TOOL_PEN,
+		.required = true,
+	},
+	[HPEN_BARREL_SWITCH] = {
+		.name = "BARL",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_BARREL_SWITCH),
+		.code = BTN_STYLUS,
+		.required = false,
+	},
+	[HPEN_INVERT] = {
+		.name = "INVR",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_INVERT),
+		.code = BTN_TOOL_RUBBER,
+		.required = true,
+	},
+	[HPEN_ERASER] = {
+		.name = "ERSR",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_ERASER),
+		.code = BTN_TOUCH,
```

Comment:
I think BTN_TOUCH is wrong here. It should be emitted at start of any action, see TIP_SWITCH.

Reply:
The protocol is a bit weird with this.

`HUD_INVERT` is the "we're using the eraser now":

> Indicates that the pen's orientation implies an intention to erase

while `HUD_ERASER` is, as I understand it, just a `TIP_SWITCH`-when-erasing:

> Indicates that the pen is erasing

I think Linux does it like that. (Argh, I lost the file where they actually handle generic HID again.)

Reply:
> as I understand it, just a TIP_SWITCH-when-erasing:
Did you check the number of /dev/input/* character devices Linux creates for the pen?
I bet it is more than one as there should be a way to separate pen events from eraser ones

Reply:
Just looked at Linux's drivers/hid/hid-input.c. They maps HUD_ERASER to BTN_TOUCH too. And allows to create combined pen/eraser devices. You are right.